chain/rpc/src/client.rs

```diff
@@ -437,31 +543,221 @@ pub mod tests {
+            .await
+            .expect_err("expected error");
+
+        m.assert();
+        assert!(matches!(err, JsonRpcProviderClientError::JsonRpcError(_)));
+    }
+
+    #[async_std::test]
+    async fn test_client_should_not_retry_on_nonretryable_json_rpc_error() {
+        let mut server = mockito::Server::new_async().await;
+
+        let m = server
+            .mock("POST", "/")
+            .with_status(200)
+            .match_body(mockito::Matcher::PartialJson(json!({"method": "eth_blockNumber"})))
+            .with_body(
+                r#"{
+              "jsonrpc": "2.0",
+              "id": 1,
+              "error": {
+                "message": "some message",
+                "code": -32000
+              }
+            }"#,
+            )
+            .expect(1)
+            .create();
+
+        let client = JsonRpcProviderClient::new(
+            &server.url(),
+            SurfRequestor::default(),
+            SimpleJsonRpcRetryPolicy {
+                max_retries: Some(2),
+                retryable_json_rpc_errors: vec![],
+                initial_backoff: Duration::from_millis(100),
+                ..SimpleJsonRpcRetryPolicy::default()
+            },
+        );
+
+        let err = client
+            .request::<_, ethers::types::U64>("eth_blockNumber", ())
+            .await
+            .expect_err("expected error");
 
-        mock_requestor
-            .expect_http_post()
-            .once()
-            .withf(|url, _| url == "localhost")
-            .returning(|_, _| futures::future::ok("}{malformed".to_string()).boxed());
+        m.assert();
+        assert!(matches!(err, JsonRpcProviderClientError::JsonRpcError(_)));
+    }
 
-        let client = JsonRpcProviderClient::new("localhost".into(), mock_requestor);
+    #[async_std::test]
+    async fn test_client_should_retry_on_malformed_json_rpc_error() {
+        let mut server = mockito::Server::new_async().await;
+
+        let m = server
+            .mock("POST", "/")
+            .with_status(200)
+            .match_body(mockito::Matcher::PartialJson(json!({"method": "eth_blockNumber"})))
+            .with_body(
+                r#"{
+              "jsonrpc": "2.0",
+              "error": {
+                "message": "some message",
+                "code": -32600
+              }
+            }"#,
+            )
+            .expect(3)
+            .create();
+
+        let client = JsonRpcProviderClient::new(
+            &server.url(),
+            SurfRequestor::default(),
+            SimpleJsonRpcRetryPolicy {
+                max_retries: Some(2),
+                retryable_json_rpc_errors: vec![-32600],
+                initial_backoff: Duration::from_millis(100),
+                ..SimpleJsonRpcRetryPolicy::default()
+            },
+        );
 
         let err = client
             .request::<_, ethers::types::U64>("eth_blockNumber", ())
             .await
             .expect_err("expected error");
+
+        m.assert();
```

Comment:
How do you know that at this point retries were made? Also a similar test which then succeeds after retries would be useful.

Reply:
See the `expect(3)` clause on L741